androguard/core/apk/__init__.py

```diff
@@ -1501,33 +1504,204 @@ def is_androidtv(self) -> bool:
+
+    @staticmethod
+    def verify_signature(signer_info, matching_certificate, signed_data, digest_algorithm):
+        matching_certificate_verified = None
+        signature = signer_info['signature'].native
+        loaded_cert = asymmetric.load_certificate(matching_certificate.chosen)
+        public_key = loaded_cert.public_key
+        key_type = loaded_cert.algorithm
+        # Verify the signature
+        try:
+            if key_type == 'rsa':
+                asymmetric.rsa_pkcs1v15_verify(public_key, signature, signed_data, hash_algorithm=digest_algorithm)
+            elif key_type == 'dsa':
+                asymmetric.dsa_verify(public_key, signature, signed_data, hash_algorithm=digest_algorithm)
+            elif key_type == 'ec':
+                asymmetric.ecdsa_verify(public_key, signature, signed_data, hash_algorithm=digest_algorithm)
+            else:
+                raise ValueError(f"Unsupported key algorithm: {key_type}")
+            matching_certificate_verified = matching_certificate.chosen.dump()
+        except errors.SignatureError:
+            logger.info(
+                f"The public key of the certificate:{hashlib.sha256(matching_certificate.chosen.dump()).hexdigest()} is not associated with the signature!")
+        return matching_certificate_verified
 
-    def get_certificate(self, filename:str) -> x509.Certificate:
+    @staticmethod
+    def get_hash_algorithm(signer_info):
+        # Determine the hash algorithm from the SignerInfo
+        digest_algorithm = signer_info['digest_algorithm']['algorithm'].native
+        # Map the digest algorithm to a hash function
+        hash_algorithms = {
+            'md5': md5,
+            'sha1': sha1,
+            'sha256': sha256,
+            'sha224': sha224,
+            'sha384': sha384,
+            'sha512': sha512
+        }
+        if digest_algorithm not in hash_algorithms:
+            raise ValueError(f"Unsupported hash algorithm: {digest_algorithm}")
+        return digest_algorithm
+
+    @staticmethod
+    def find_certificate(signed_data_certificates, signer_info):
+        """
+        From the bag of certs, obtain the certificate referenced by the SignerInfo.
+
+        Args:
+            signed_data_certificates: List of certificates in the SignedData.
+            signer_info: SignerInfo object containing the issuer and serial number reference.
+
+        Returns:
+            The matching certificate if found, otherwise None.
+        """
+        matching_certificate = None
+        issuer_and_serial_number = signer_info['sid'].native
+        issuer = issuer_and_serial_number['issuer']
+        serial_number = issuer_and_serial_number['serial_number']
+
+        # Create a x509.Name object for the issuer in the SignerInfo
+        issuer_name = x509.Name.build(issuer)
+        issuer_str = issuer_name.human_friendly
+        issuer_str = unicodedata.normalize('NFKD', issuer_str.upper().lower())
```

Comment:
Nice. This seems like a pretty neat way to handle case and unicode normalisation. I suspect there may be some edge cases but at least I can't think of a way to break it easily myself.

Unfortunately, that still leaves the fact that `.human_friendly` doesn't handle ordering correctly for our purposes. And that I can easily break to have this code produce a different result from `apksigner`/Android.

Reply:
Also, it seems `.human_friendly` does not normalise whitespace at all.

Reply:
it seems that the issuer from `CertificateChoices` and from `SignerInfo` return an ordered dictionary. Could you please provide me a dummy apk to test this against and see where it fails?

Reply:
```sh
openssl req -x509 -newkey rsa:2048 -sha512 -outform DER -out cert-rsa-1.der -days 10000 -nodes -subj '/CN= foo+CN=bar /OU=a/OU=b/C=xx' -keyout - | openssl pkcs8 -topk8 -nocrypt -outform DER -out privkey-rsa-1.der
openssl req -x509 -newkey rsa:2048 -sha512 -outform DER -out cert-rsa-2.der -days 10000 -nodes -subj '/C=xx/CN=bar+CN=foo/OU=b/OU=a' -keyout - | openssl pkcs8 -topk8 -nocrypt -outform DER -out privkey-rsa-2.der
```

```groovy
// javacert.groovy
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
def i = new FileInputStream(args[0])
def cert = CertificateFactory.getInstance("X.509").generateCertificate(i)
println cert.getIssuerX500Principal().getName(javax.security.auth.x500.X500Principal.CANONICAL)
```

```py
# javacert.py
import sys
from asn1crypto import x509
d = open(sys.argv[1], "rb").read()
c = x509.Certificate.load(d)
n = c["tbs_certificate"]["issuer"]
print(n.native)
print(n.human_friendly)
```

```
$ groovy javacert.groovy cert-rsa-1.der
c=xx,ou=b,ou=a,cn=bar+cn=foo
$ groovy javacert.groovy cert-rsa-2.der
ou=a,ou=b,cn=bar+cn=foo,c=xx
$ py3 javacert.py cert-rsa-1.der
OrderedDict([('common_name', [' foo', 'bar ']), ('organizational_unit_name', ['a', 'b']), ('country_name', 'xx')])
Common Name: bar ,  foo; Organizational Unit: b, a; Country: xx
$ py3 javacert.py cert-rsa-2.der
OrderedDict([('country_name', 'xx'), ('common_name', ['bar', 'foo']), ('organizational_unit_name', ['b',
'a'])])
Organizational Unit: a, b; Common Name: foo, bar; Country: xx
```

Reply:
For extra fun:

```
$ openssl req -x509 -newkey rsa:2048 -sha512 -outform DER -out cert-rsa-3.der -days 10000 -nodes -utf8 -subj '/CN=x/CN=foo  +CN=Ii   +CN=İı +CN=ẞß+CN=bar  /CN=y' -keyout - | openssl pkcs8 -topk8 -nocrypt -outform DER -out privkey-rsa-3.der
$ sed 's/bar/qux/' < cert-rsa-3.der > cert-rsa-3_.der
$ groovy javacert.groovy cert-rsa-3.der
cn=y,cn=bar+cn=foo+cn=ii+cn=ii+cn=ßss,cn=x
$ groovy javacert.groovy cert-rsa-3_.der
cn=y,cn=foo+cn=ii+cn=ii+cn=qux+cn=ßss,cn=x
$ py3 javacert.py cert-rsa-3.der
OrderedDict([('common_name', ['x', 'Ii   ', 'bar  ', 'foo  ', 'İı ', 'ẞß', 'y'])])
Common Name: y, ẞß, İı , foo  , bar  , Ii   , x
common name: y, ßss, ii , foo  , bar  , ii   , x
$ py3 javacert.py cert-rsa-3_.der
OrderedDict([('common_name', ['x', 'Ii   ', 'qux  ', 'foo  ', 'İı ', 'ẞß', 'y'])])
Common Name: y, ẞß, İı , foo  , qux  , Ii   , x
common name: y, ßss, ii , foo  , qux  , ii   , x
```

Reply:
This gets pretty close: https://gist.github.com/obfusk/b337e4f641991cc88a24d3f9567db80f

Reply:
(for comparison the abbreviations aren't needed, that was just to make it easier to compare the output)

Reply:
> This gets pretty close: https://gist.github.com/obfusk/b337e4f641991cc88a24d3f9567db80f

Improved it more. And found a difference between android and openjdk (and thus `apksigner`). But it should be really close now. I can't guaranteed there are no further edge cases, but I don't know how to break it :P

Reply:
script to generate test cert:

```sh
openssl req -x509 -quiet -newkey rsa:2048 -sha512 -outform DER -out cert-rsa-3.der -days 10000 -nodes -utf8 -config openssl.cnf -subj '/custom2=42+custom1=37+O=org/emailAddress=x@y/CN= x '$'\t\t'' 猫x/CN=foo  +CN=Ii   +CN=İı +CN=ẞß+CN=bar  +CN=zz+CN= #,;\+\\/CN=y' -keyout - | openssl pkcs8 -topk8 -nocrypt -outform DER -out privkey-rsa-3.der
```

openssl.cnf:

```
oid_section = myoids

[ myoids ]
custom1=0.1.2.3
custom2=0.1.11.3
```